File: back-end/internal/chat/chatapplication/chat_service.go
// internal/chat/chatapplication/chat_service.go
package chatapplication

import (
	"context"
	"fmt"

	"back-end/internal/chat/chatdomain"
	"back-end/internal/chat/chatinfrastructure"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// ChatService se encarga de la lógica de negocio del chat.
type ChatService struct {
	ChatRepo *chatinfrastructure.ChatRepository
}

// NewChatService crea una nueva instancia de ChatService.
func NewChatService(chatRepo *chatinfrastructure.ChatRepository) *ChatService {
	return &ChatService{
		ChatRepo: chatRepo,
	}
}

// SendMessage envía un mensaje entre dos usuarios.
// Se espera que en msg se tenga senderId, receiverId y text.
func (s *ChatService) SendMessage(ctx context.Context, msg chatdomain.ChatMessage) (chatdomain.ChatMessage, error) {
	// Ejemplo de validación: Si el chat está bloqueado por el emisor, se rechaza el envío.
	// (La verificación se puede hacer obteniendo el ChatRoom y comprobando si msg.SenderID se encuentra en BlockedBy).
	// Aquí se asume que dicha validación se implementa en el repositorio o previamente.

	// Para notificaciones, se define el nombre del remitente (este ejemplo es simple; ajusta según convenga).
	senderName := "Remitente" // O bien, obtenerlo de la base de datos

	return s.ChatRepo.SendMessage(ctx, msg, senderName)
}

// GetMessagesBetween obtiene los mensajes intercambiados entre dos usuarios.
func (s *ChatService) GetMessagesBetween(ctx context.Context, user1, user2 string) ([]chatdomain.ChatMessage, error) {
	return s.ChatRepo.GetMessagesBetween(ctx, user1, user2)
}

// MarkMessageAsRead marca un mensaje como leído.
func (s *ChatService) MarkMessageAsRead(ctx context.Context, messageID string) error {
	return s.ChatRepo.MarkMessageAsRead(ctx, messageID)
}

// BlockUser en un chat, agregando al usuario que bloquea.
func (s *ChatService) BlockUser(ctx context.Context, chatRoomID, blockerID string) error {
	return s.ChatRepo.BlockUser(ctx, chatRoomID, blockerID)
}

// GetChatRoom obtiene o crea un ChatRoom entre dos usuarios.
func (s *ChatService) GetChatRoom(ctx context.Context, userId, partnerId string) (chatdomain.ChatRoom, error) {
	u1, err := primitive.ObjectIDFromHex(userId)
	if err != nil {
		return chatdomain.ChatRoom{}, err
	}
	u2, err := primitive.ObjectIDFromHex(partnerId)
	if err != nil {
		return chatdomain.ChatRoom{}, err
	}
	return s.ChatRepo.GetOrCreateChatRoom(ctx, u1, u2)
}

// internal/chat/chatapplication/chat_service.go

// GetChatRooms obtiene los chats del usuario, paginados.
// page comienza en 1. skip = (page - 1) * limit.
func (s *ChatService) GetChatRooms(ctx context.Context, userID string, limit, page int) ([]chatdomain.ChatDetails, error) {
	uID, err := primitive.ObjectIDFromHex(userID)
	if err != nil {
		return nil, fmt.Errorf("userID inválido: %v", err)
	}
	skip := (page - 1) * limit
	return s.ChatRepo.GetChatRooms(ctx, uID, limit, skip)
}

File: front-end/services/chatService.ts
import Constants from "expo-constants";
const API = "http://192.168.0.28:9000";


export interface ChatRoom {
    id: string;
    participant1: string;
    participant2: string;
    blockedBy?: string[];
    createdAt: string;
    updatedAt: string;
}

export interface Message {
    id: string;
    senderId: string;
    receiverId: string;
    text: string;
    createdAt: string;
}

// Obtiene o crea el chat room entre dos usuarios.
export const getChatRoom = async (
    userId: string,
    partnerId: string,
    token: string
): Promise<ChatRoom | undefined> => {
    try {
        const res = await fetch(`${API}/chat/room?user=${userId}&partner=${partnerId}`, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                "Authorization": `Bearer ${token}`
            }
        });
        if (!res.ok) {
            throw new Error(`Error HTTP: ${res.status}`);
        }
        return await res.json();
    } catch (error) {
        console.error("Error obteniendo chat room:", error);
    }
};

// Envía un mensaje de chat.
export const sendChatMessage = async (
    message: { senderId: string; receiverId: string; text: string },
    token: string
): Promise<{ data: any } | undefined> => {
    try {
        const res = await fetch(`${API}/chat/messages`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "Authorization": `Bearer ${token}`
            },
            body: JSON.stringify(message)
        });
        return await res.json();
    } catch (error) {
        console.error("Error sending chat message:", error);
    }
};

// Obtiene los mensajes entre dos usuarios (el backend utiliza internamente el ChatRoom).
export const getMessagesBetween = async (
    user1: string,
    user2: string,
    token: string
): Promise<Message[] | undefined> => {
    try {
        const res = await fetch(`${API}/chat/messages?user1=${user1}&user2=${user2}`, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                ...(token && { "Authorization": `Bearer ${token}` })
            }
        });
        if (!res.ok) {
            throw new Error(`Error HTTP: ${res.status}`);
        }
        return await res.json();
    } catch (error) {
        console.error("Error getting chat messages:", error);
    }
};

// Marca un mensaje como leído.
export const markMessageAsRead = async (
    messageId: string,
    token: string
): Promise<{ data: any } | undefined> => {
    try {
        const res = await fetch(`${API}/chat/messages/${messageId}/read`, {
            method: "POST",
            headers: {
                "Content-Type": "application/json",
                "Authorization": `Bearer ${token}`
            }
        });
        if (!res.ok) {
            throw new Error(`Error HTTP: ${res.status}`);
        }
        return await res.json();
    } catch (error) {
        console.error("Error marking message as read:", error);
    }
};
// Obtiene los chats del usuario autenticado, paginados de 10 en 10.
export const getChatRooms = async (
    token: string,
    limit: number = 10,
    page: number = 1
): Promise<ChatRoom[] | undefined> => {
    try {
        const res = await fetch(`${API}/chat/rooms?limit=${limit}&page=${page}`, {
            method: "GET",
            headers: {
                "Content-Type": "application/json",
                "Authorization": `Bearer ${token}`
            }
        });

        if (!res.ok) {
            throw new Error(`Error HTTP: ${res.status}`);
        }
        return await res.json();
    } catch (error) {
        console.error("Error obteniendo chat rooms:", error);
    }
};
